### PR TITLE
fix node failover. add proper tests for unreachable node and maxconne…

### PR DESCRIPTION
**Pull request by christophermancini**
…ctions.

To address issue #113 (CLIENTS-700)

### CODE REVIEW

**Comment by lukebakken** [comment]
Travis is :cry:

File: src/Riak.php [hunk 1]
```diff
@@ -213,10 +226,15 @@ public function getLastRequest()
...
      * @throws Exception
      */
-    protected function pickNewNode()
+    public function pickNewNode()
     {
-        // mark current active node as inactive
+        // mark current active node as inactive and increment attempts
         $this->getActiveNode()->setInactive(true);
+        $this->attempts++;
+        $this->inactiveNodes[$this->getActiveNodeIndex()] = $this->getActiveNode();
```
Comment (lukebakken) [inline src/Riak.php]: How does a node become active again?

Comment (christophermancini) [inline src/Riak.php]: I didn't add that functionality because PHP doesn't run as a service, after it has completed execution the memory is cleared and starts fresh. There are probably only a few edge cases where that would be of value. I could add a method to allow the developer to invoke the inactiveNodes being merged back into the pool. Not sure if it would add much value though.

Comment (lukebakken) [inline src/Riak.php]: > the memory is cleared and starts fresh

Does this mean that connections made to Riak during one HTTP request / response cycle get torn down, every time?

Comment (christophermancini) [inline src/Riak.php]: The cURL extension will attempt to reuse connections automatically and for the PB lib I am using persistent sockets, so if they are using PHP on a webserver, there will be some performance boosts there. On the CLI side, I am pretty sure that there is no way to reuse connections unless I create a local proxy with another language that can stay open listening for requests from PHP. 

Even with that said, the reuse by cURL and sockets is only available for the same process of PHP, e.g. FastCGI runs a master process and several child processes, the reuse will work only for that child process.

Comment (lukebakken) [inline src/Riak.php]: OK thanks for the info. I read [this doc](http://php.net/manual/en/features.persistent-connections.php) to get a bit more information about how PHP works.

Is there any way to persist node state across requests, so you can remember the inactive nodes and not re-try on the next HTTP request? I envision a scenario where quite a bit of time is lost sending requests to a persistently-down node. Of course, this requires a method to detect when a node has recovered, too. Just something to think about.

Comment (christophermancini) [inline src/Riak.php]: Without using a third party library to add the ability to share memory between PHP processes or sessions, I don't think it would be beneficial. It would be best for the developer to determine if that state data is relevant to them.

Comment (lukebakken) [inline src/Riak.php]: It wouldn't be for developer use, but for performance. For instance, if you have 5 nodes and one is down, there is a 20% chance of selecting the down node for an operation, trying to execute the operation and losing time through that entire process before selecting a different node. Over the duration of time that the node is down that time adds up fast.

File: composer.json [hunk 2]
```diff
@@ -58,7 +58,7 @@
     "ext-curl": "*"
   },
   "require-dev": {
-    "phpunit/phpunit": "4.5.*",
+    "phpunit/phpunit": "4.8.*",
```
Comment (debo) [inline composer.json]: @christophermancini this should probably be part of a different PR (just being picky as usual)

File: src/Riak.php (hunk 1 above)
Comment (debo) [inline src/Riak.php]: @lukebakken how is this generally handled in other clients?

Comment (lukebakken) [inline src/Riak.php]: Node errors or failures in health check will mark a node as offline. Health check runs periodically to test offline nodes, and if they are found to be online, returns them to the set of running nodes.

This requires the ability to save node state which could be accomplished via shared memory and semaphores in PHP. I may put together a test to see how much a down node affects the PHP client.

File: composer.json (hunk 2 above)
Comment (christophermancini) [inline composer.json]: Yea, you are right, I originally made the change because I was trying to fix the flip flopping test, thinking that maybe there was a bug... and then I ended up changing that test to be skipped in all cases anyways. If this were a non "-dev" dependency, I wouldn't have bundled it together.

File: src/Riak.php (hunk 1 above)
Comment (debo) [inline src/Riak.php]: The problem is that none of these feature are natively supported. You have to specifically compile PHP enabling the right extension which might be difficult for most users.

Comment (lukebakken) [inline src/Riak.php]: They're "natively" supported, just not enabled OOTB. Keeping track of down nodes may be enough of a perf benefit to make it an option to users. Only benchmarking will tell.

Comment (debo) [inline src/Riak.php]: In which case wouldn't then be better to let that job to a more suited tool? A load balancer maybe or something similar?
